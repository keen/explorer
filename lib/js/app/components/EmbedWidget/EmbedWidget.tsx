import React, { FC, useMemo, useContext } from 'react';
import { Button } from '@keen.io/ui-core';
import Prism from 'prismjs';

import { Code } from './EmbedWidget.styles';
import { createCodeSnippet } from './utils';
import text from './text.json';

import { AppContext } from '../../contexts';
import { copyToClipboard } from '../../utils';

type Props = {
  /** Widget type */
  widget: string;
  /** Query definition */
  query: Record<string, any>;
};

const EmbedWidget: FC<Props> = ({ widget, query }) => {
  const { keenAnalysis } = useContext(AppContext);
  const code = useMemo(
    () =>
      createCodeSnippet({
        widget,
        query,
        projectId: keenAnalysis.config.projectId,
        readKey: keenAnalysis.config.readKey,
      }),
    [widget, query]
  );

  const html = useMemo(
    () => Prism.highlight(code, Prism.languages.javascript, 'javascript'),
    [code]
  );

  return (
    <div>
      <Code dangerouslySetInnerHTML={{ __html: html }} />
      <Button
        variant="secondary"
        style="outline"
        onClick={() => copyToClipboard(code)}
      >
        {text.copyLabel}
      </Button>
    </div>
  );
};

export default EmbedWidget;
