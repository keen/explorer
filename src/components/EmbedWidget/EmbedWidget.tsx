import React, { FC } from 'react';
import { Light as SyntaxHighlighter } from 'react-syntax-highlighter';
import xml from 'react-syntax-highlighter/dist/cjs/languages/hljs/xml';
import js from 'react-syntax-highlighter/dist/cjs/languages/hljs/javascript';
import a11y from 'react-syntax-highlighter/dist/cjs/styles/hljs/a11y-light';

SyntaxHighlighter.registerLanguage('xml', xml);
SyntaxHighlighter.registerLanguage('javascript', js);

type Props = {
  children: string;
};

const EmbedWidget: FC<Props> = ({ children }) => (
  <SyntaxHighlighter
    language="javascript"
    style={a11y}
    wrapLongLines={true}
    customStyle={{ margin: '0' }}
  >
    {children}
  </SyntaxHighlighter>
);

export default EmbedWidget;
