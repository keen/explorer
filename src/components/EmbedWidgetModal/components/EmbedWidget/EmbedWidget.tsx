import React, { useCallback, useContext, useMemo } from 'react';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';

import { Button, ModalHeader } from '@keen.io/ui-core';
import { BodyText } from '@keen.io/typography';
import { colors } from '@keen.io/colors';

import { AppContext } from '../../../../contexts';
import { appActions, appSelectors } from '../../../../modules/app';
import { getQuerySettings } from '../../../../modules/queries';
import { createBodyCode, createHeadCode } from '../../../../utils';

import {
  EmbedLabel,
  ModalBody,
  ModalWrapper,
  Navigation,
  NavigationItem,
  Section,
  StepLabel,
  StepNumber,
} from '../../EmbedWidgetModal.styles';

import { Syntax } from '../index';

const EmbedWidget = () => {
  const { t } = useTranslation(null, { useSuspense: false });
  const dispatch = useDispatch();
  const { keenAnalysis, datavizSettings } = useContext(AppContext);
  const { projectId, readKey } = keenAnalysis.config;
  const theme = datavizSettings.theme;

  const { type: widget, widgetSettings, chartSettings } = useSelector(
    appSelectors.getVisualization
  );
  const query = useSelector(getQuerySettings);

  const headCode = useMemo(() => createHeadCode(), []);

  const bodyCode = useMemo(
    () =>
      createBodyCode({
        widget,
        query,
        chartSettings: { ...chartSettings, theme },
        widgetSettings,
        projectId,
        readKey,
      }),
    []
  );

  const closeHandler = useCallback(
    () => dispatch(appActions.hideEmbedModal()),
    []
  );

  return (
    <ModalWrapper>
      <ModalHeader onClose={closeHandler}>
        {t('embed_widget.embed_html_title')}
      </ModalHeader>
      <ModalBody>
        <EmbedLabel>
          <BodyText
            variant="body2"
            color={colors.green[500]}
            fontWeight={'bold'}
          >
            {t('embed_widget.embed_title')}
          </BodyText>
        </EmbedLabel>
        <BodyText variant="body2">{t('embed_widget.embed_text')}</BodyText>
        <Section>
          <StepLabel>
            <StepNumber>
              <BodyText
                variant="body2"
                color={colors.green[500]}
                fontWeight={'bold'}
              >
                {t('embed_widget.embed_step_label')} 1
              </BodyText>
            </StepNumber>
            <BodyText
              variant="body2"
              dangerouslySetInnerHTML={{
                __html: t('embed_widget.embed_step_first', {
                  tag: '<strong>&lt;head&gt;</strong>',
                  interpolation: { escapeValue: false },
                }),
              }}
            ></BodyText>
          </StepLabel>
          <Syntax code={headCode} />
        </Section>
        <Section>
          <StepLabel>
            <StepNumber>
              <BodyText
                variant="body2"
                color={colors.green[500]}
                fontWeight={'bold'}
              >
                {t('embed_widget.embed_step_label')} 2
              </BodyText>
            </StepNumber>
            <BodyText
              variant="body2"
              dangerouslySetInnerHTML={{
                __html: t('embed_widget.embed_step_second', {
                  tag: '<strong>&lt;body&gt;</strong>',
                  interpolation: { escapeValue: false },
                }),
              }}
            />
          </StepLabel>
          <Syntax code={bodyCode} />
        </Section>
      </ModalBody>
      <Navigation>
        <NavigationItem>
          <BodyText variant="body2">
            {t('embed_widget.do_you_need_full_page_code')}
          </BodyText>
        </NavigationItem>
        <NavigationItem>
          <Button
            variant="secondary"
            style="outline"
            onClick={() =>
              dispatch(appActions.downloadCodeSnippet(projectId, readKey))
            }
          >
            {t('embed_widget.download_file_button')}
          </Button>
        </NavigationItem>
      </Navigation>
    </ModalWrapper>
  );
};

export default EmbedWidget;
