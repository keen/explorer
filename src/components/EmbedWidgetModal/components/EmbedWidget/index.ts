import EmbedWidget from './EmbedWidget';

export default EmbedWidget;
