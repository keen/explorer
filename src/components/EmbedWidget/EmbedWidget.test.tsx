import React from 'react';
import { render } from '@testing-library/react';

import EmbedWidget from './EmbedWidget';

test('renders code snippet', () => {
  const code = '<script>() => {}</script>';
  const { container } = render(<EmbedWidget>{code}</EmbedWidget>);

  expect(container).toMatchSnapshot();
});
