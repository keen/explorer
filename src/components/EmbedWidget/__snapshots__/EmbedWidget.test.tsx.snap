// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders code snippet 1`] = `
<div>
  <pre
    style="display: block; overflow-x: auto; background: rgb(254, 254, 254); color: rgb(84, 84, 84); padding: 0.5em; margin: 0px;"
  >
    <code
      class="language-javascript"
      style="white-space: pre-wrap;"
    >
      <span>
        <span
          class="comment linenumber react-syntax-highlighter-line-number"
          style="display: inline-block; min-width: 1em; padding-right: 1em; text-align: right; user-select: none;"
        >
          1
        </span>
        <span>
          &lt;script&gt;
        </span>
        <span
          class="hljs-function"
        >
          () =&gt;
        </span>
        <span>
           {}&lt;/script&gt;
        </span>
      </span>
    </code>
  </pre>
</div>
`;
